simple_network_sim/inference.py

```diff
@@ -0,0 +1,702 @@
+            shape: float,
+            kernel_sigma: float,
+            random_state: np.random.Generator
+    ):
+        self.value = value
+        self.mean = mean
+        self.shape = shape
+        self.kernel_sigma = kernel_sigma
+        self.random_state = random_state
+
+    @staticmethod
+    def generate_from_prior(fitter: ABCSMC) -> InferredInfectionProbability:
+        """ Sample from prior distribution. For infection probability, we use
+        the value in the data pipeline as mean of our prior, and shape parameter
+        is fixed at 2. This defines a Beta distribution centered around our prior.
+
+        :param fitter: ABC-SMC fitter object
+        :return: New InferredInitialInfections randomly sampled from prior
+        """
+        shape = fitter.infection_probability_shape
+        kernel_sigma = fitter.infection_probability_kernel_sigma
+        mean = fitter.infection_probability.at[0, "Value"]
+        value = fitter.infection_probability.copy().assign(
+            Value=lambda x: stats.beta.rvs(shape, shape * (1 - x.Value) / x.Value, random_state=fitter.random_state))
+        return InferredInfectionProbability(value, mean, shape, kernel_sigma, fitter.random_state)
+
+    def generate_perturbated(self) -> InferredInfectionProbability:
+        """ From current parameter, add a perturbation to infection probability and return
+        a newly created perturbated parameter:
+
+        .. math::
+            P_t^* ~ K(P_t | P_{t-1}) ~ P_{t-1} + Uniform([-1,1]) * kernel_sigma
+
+        :return: New parameter which is similar to self up to a perturbation
+        """
+        value = self.value.copy()
+        value.Value = value.Value.apply(lambda x: x + self.kernel_sigma * stats.uniform
+                                        .rvs(-1., 2., random_state=self.random_state))
+        return InferredInfectionProbability(value, self.mean, self.shape, self.kernel_sigma, self.random_state)
+
+    def validate(self) -> bool:
+        """ Checks that the particle is valid, i.e. that infection probability is
+        between 0 and 1.
+
+        :return: Whether the parameter is valid
+        """
+        return np.all(self.value.Value > 0.) and np.all(self.value.Value < 1.)
+
+    def prior_pdf(self) -> float:
+        """ Compute pdf of the prior distribution evaluated at the parameter x.
+        Infection probability has a prior a beta distribution. The pdf is evaluated
+        at the current value of the parameter.
+
+        :return: pdf value of prior distribution evaluated at x
+        """
+        return stats.beta.pdf(self.value.at[0, "Value"], self.shape, self.shape * (1 - self.mean) / self.mean)
+
+    def perturbation_pdf(self, x: pd.DataFrame) -> float:
+        """ Compute pdf of the perturbation evaluated at the parameter x,
+        from the current parameter. In ABC-SMC when a particle is sampled
+        from the previous population it is slightly perturbed:
+
+        .. math::
```

Comment:
I think you need an empty line between `.. math::`  and the formula

Reply:
Thanks. Is there any way for me to check this before it's pushed? Also do I need to start the docstring with r""" as I say in _init_.py?

Reply:
You can do a local build of the readthedocs site.

Reply:
https://github.com/ScottishCovidResponse/simple_network_sim#building-the-documentation

Reply:
thanks!